src/daemon/osdkeys.vala

```diff
@@ -0,0 +1,254 @@
+			mixer.default_sink_changed.connect(on_mixer_sink_changed);
+			mixer.open();
+
+			// wait a short while to allow the async mixer methods to complete otherwise
+			// we see the volume OSD on startup accidently when keymap changes are invoked
+			Timeout.add(200, () => {
+				initialising = false;
+				map = Gdk.Keymap.get_for_display(Gdk.Display.get_default());
+				map.state_changed.connect(on_keymap_state_changed);
+				return false;
+			});
+		}
+
+		void on_mixer_sink_changed(uint id) {
+			set_default_mixer();
+		}
+
+		void set_default_mixer() {
+			if (stream != null) {
+				SignalHandler.disconnect(stream, notify_id);
+			}
+
+			stream = mixer.get_default_sink();
+			notify_id = stream.notify.connect(on_stream_notify);
+			update_volume();
+		}
+
+		void on_stream_notify(Object? o, ParamSpec? p) {
+			if (p.name == "volume" || p.name == "is-muted") {
+				update_volume();
+			}
+		}
+
+		/**
+		 * Called when something changes on the mixer, i.e. we connected
+		 * This is where we hook into the stream for changes
+		 */
+		void on_mixer_state_change(uint new_state) {
+			if (new_state == Gvc.MixerControlState.READY) {
+				set_default_mixer();
+			}
+		}
+
+		/**
+		* Update the OSD when something changes (volume/mute)
+		*/
+		void update_volume() {
+			if (initialising) return;
+
+			var vol_norm = mixer.get_vol_max_norm();
+			var vol = stream.get_volume();
+
+			/* Same maths as computed by volume.js in gnome-shell, carried over
+			* from C->Vala port of budgie-panel */
+			int n = (int) Math.floor(3*vol/vol_norm)+1;
+			string image_name;
+
+			// Work out an icon
+			if (stream.get_is_muted() || vol <= 0) {
+				image_name = "audio-volume-muted-symbolic";
+			} else {
+				switch (n) {
+					case 1:
+						image_name = "audio-volume-low-symbolic";
+						break;
+					case 2:
+						image_name = "audio-volume-medium-symbolic";
+						break;
+					default:
+						image_name = "audio-volume-high-symbolic";
+						break;
+				}
+			}
+
+			var pct = (float)vol / (float)vol_norm;
+			if (pct > 1.0 && !stream.get_is_muted()) {
+				image_name = "audio-volume-overamplified-symbolic";
+			}
+
+			HashTable<string,Variant> params;
+			params = new HashTable<string,Variant>(null, null);
+			params.set("level", new Variant.double(pct));
+			params.set("icon", new Variant.string(image_name));
+
+			osd.ShowOSD.begin(params);
+		}
+
+		/*
+		  we allow Gtk keymap to control when to display capslock/numlock state
+		  It does appear though that there is a bug under wayland where the
+		  state of the capslock/numlock is not set until another keymap is activated
+		  This can be ctrl/shift as well as the caps/numlock. If the latter is activated
+		  this can mean the OSD for keymap state is not activated on first use; only on
+		  subsequent keypresses.
+		 */
+		private void on_keymap_state_changed() {
+			if (!firstrun) {
+				capslock = map.get_caps_lock_state();
+				numlock = map.get_num_lock_state();
+				firstrun = true;
```

Comment:
This feels backwards. Doesn't this actually happen on the first run?

Reply:
As per the comment above - 'fraid not.  It appears GTK3 is confused under wayland with states and it needs the user to hit ctrl/shift/caps/numlock for it to update the states correctly.